Mortgage Calculator release

ErazorWhite opened the pull request.

scd-hoi committed:
removed trash

scd-hoi committed:
added modern-normalize, color vars, reset h-ul-p styles

scd-hoi committed:
installed normalize & styled

scd-hoi committed:
added fonts

scd-hoi committed:
Migrating to TS

scd-hoi committed:
font fix

scd-hoi committed:
logic implemented, still work in progress

scd-hoi committed:
yup & @hookform/resolvers installed

scd-hoi committed:
CalculatorForm in progress

scd-hoi committed:
Refactoring

scd-hoi committed:
installed linaria, postCss custom media

scd-hoi committed:
added custom media breakpoints

scd-hoi committed:
Container added

scd-hoi committed:
changed linaria plugins installation order

scd-hoi committed:
tsconfig simplified

scd-hoi committed:
linaria removed, styled-components installed instead

scd-hoi committed:
breakpoints moved from custom global postcss to js constants

scd-hoi committed:
moved to styled-components & new breakpoints

scd-hoi committed:
fonts presets created in typography

scd-hoi committed:
button styles reset

scd-hoi committed:
added opacity lime color

scd-hoi committed:
Input reworked, added styles

scd-hoi committed:
Added styles

scd-hoi committed:
removed unnecessary comment

scd-hoi committed:
ClearAll fix

scd-hoi committed:
FC, ReactNode props interface renamed and moved to types

scd-hoi committed:
IconCalculator added

scd-hoi committed:
color-black .25 added

scd-hoi committed:
unnecessary icon removed

scd-hoi committed:
padding & key fixes

scd-hoi committed:
unnecessary icon removed [2]

scd-hoi committed:
ResultsBackground svg added

scd-hoi committed:
CalcButton & Results update

scd-hoi committed:
Added Intl.NumberFormat

scd-hoi committed:
clearButton was made more clear & mode: "onTouched" enabled

scd-hoi committed:
reworked to remove :has()

scd-hoi committed:
big update for inputs

scd-hoi committed:
icons moved to components

scd-hoi committed:
radio button span changed to ::before, lime-opacity fixed

scd-hoi committed:
padding changed to margin

scd-hoi committed:
icon added

scd-hoi committed:
radiobutton checked border color fix

scd-hoi committed:
text inputs & breakpoints update

scd-hoi committed:
just another input update

scd-hoi committed:
added desktop styles, types & a bit of refactoring

scd-hoi committed:
say goodbye to styled props

scd-hoi committed:
container reworked

scd-hoi committed:
validation styles implemented, input value issue fixed

scd-hoi committed:
more accurately specified animation var & added new lime

scd-hoi committed:
hover & focus handlers added

scd-hoi committed:
node.js CI added

scd-hoi committed:
tsconfig path fix

scd-hoi committed:
CI fix pnpm added to deployment

scd-hoi committed:
update base path vite config

scd-hoi committed:
data-attr changed to styled props for validation

scd-hoi committed:
base path vite config fix

scd-hoi committed:
html path fix

scd-hoi committed:
svg icon color fix

scd-hoi committed:
fonts update + line-height fix

scd-hoi committed:
container 100vh removed, body padding-top added

scd-hoi committed:
input transition animation added

scd-hoi committed:
postcss removed

scd-hoi committed:
padding-top update

scd-hoi committed:
README.md update, .PNGs added

scd-hoi committed:
just another padding-top little fix

scd-hoi committed:
font color fix

scd-hoi committed:
box-shadow added to form

scd-hoi committed:
desktop padding added

scd-hoi committed:
removed empty row

scd-hoi committed:
BP renaming

afitiskin commented on src/assets/icons/icon.svg:
Если это favicon, то желательно его так и назвать, favicon.svg

ErazorWhite replied:
done

afitiskin commented on src/components/CalculateButton/CalculateButton.styled.ts:
```diff
@@ -0,0 +1,30 @@
+    }
+`
+
+export const P = styled.p`
```
это тут откуда и зачем?

ErazorWhite replied:
Старье какое-то, убиремс

afitiskin commented on src/components/CalculateButton/CalculateButton.styled.ts:
```diff
@@ -0,0 +1,30 @@
+    color: var(--color-state-900);
+    fill: var(--color-state-900);
+    border: none;
+    border-radius: 999px;
```
мне кажется 50% будет достаточно

ErazorWhite replied:
В макете так дизайнер указал
![image](https://github.com/user-attachments/assets/053c5f93-b854-458b-a688-abb484e30805)

Если взять 50%, то это уже перебор
![image](https://github.com/user-attachments/assets/7f210fbe-b6ae-4e80-8a3a-e9411337cd12)

В итоге 30px выглядят также как и 999px. Оставлю 30px.

afitiskin replied:
Ок, пусть будет как было или 30px, я забыл что 50% только на квадратах работает

afitiskin commented on src/components/CalculateButton/CalculateButton.tsx:
```diff
@@ -0,0 +1,13 @@
+    return (
+        <StyledButton>
+            <IconCalculator/>
+            <P>{children}</P>
```
Абзац внутри кнопки? Ничего лучше не нашлось? Он вообще нужен тут?

ErazorWhite replied:
Прибрал

afitiskin commented on src/components/CalculateButton/CalculateButton.tsx:
```diff
@@ -0,0 +1,13 @@
+
+export const CalculateButton: FC<Props> = ({children}) => {
+    return (
+        <StyledButton>
```
Ну допустим я хочу задать кнопке tabindex, или type, или aria- свойства например, что мне делать?

ErazorWhite replied:
решается через распыление {...props}
поправил

afitiskin replied:
Ну распыление ...props без правильного описания этих props выдаст тебе ошибку при попытке эти параметры добавить `<CalculateButton  type="button">` будет ругаться на то, что нет такого типа у этой компоненты

ErazorWhite replied:
та не, типизация сама подтянулась
<img width="857" alt="image" src="https://github.com/user-attachments/assets/341b7512-51e5-495f-a595-24de8078898a">
я ж в самом компоненте типизацию описываю
<img width="743" alt="image" src="https://github.com/user-attachments/assets/9cfadadc-d9b9-4f4b-ab5b-1a0693fc10af">

afitiskin commented on src/components/Results/CalculatedResults/CalculatedResults.styled.ts:
```diff
@@ -0,0 +1,60 @@
+import {DESKTOP_BP, TABLET_BP} from "../../../global/breakpoints";
+import {ResultsBox} from "../ResultsCommon.styles";
+
+export const ResultsBoxCalculated = styled(ResultsBox)`
```
1 компонент = 1 файл, желательно. И названия вместо H2, P (если это не общепринятые на всем сайте h2 и p) лучше задавать тематические. В крайнем случае, ты можешь задать стили заголовкам и абзацам внутри контейнера:

```
const SomeContainer = styled.div`
  /* container styles */
  h2 {
    /* h2 styles */
  }
  p {
    /* p styles */
  }
`;
```

и использовать стандартные тэги чтобы не плодить лишнего

ErazorWhite replied:
Done
Соединил EmptyResults и CalculatedResults в Results.

afitiskin commented on src/components/Results/CalculatedResults/CalculatedResults.styled.ts:
```diff
@@ -0,0 +1,60 @@
+    margin-bottom: 16px;
+`
+
+export const P = styled.p`
```
Не забывай, что стили текста внутри p и без p должны быть одинаковые. Не всегда можно и имеет смысл разделять текст на абзацы, но стили при это должны совпадать

ErazorWhite replied:
Ты, наверное, имеешь ввиду, что текст внутри некоего общего контейнера не должен отличаться от такого же текста, но обёрнутого в тег `p`.  В макете пять разных пресетов настройки текста (размер, жирность, высота линии, межбуквенный интервал). Я могу задать общую настройку на весь документ через body {}. Но все равно где-то не попаду в пресет и нужно будет оборачивать в тег со своими доп. стилями текста.

afitiskin commented on src/components/Results/CalculatedResults/CalculatedResults.styled.ts:
```diff
@@ -0,0 +1,60 @@
+    ${typography.preset2};
+    color: var(--color-white);
+`
+export const CalculatedBox = styled.div`
```
Как не заглядывая в код понять для чего `ResultsBoxCalculated` а для чего `CalculatedBox` и чем они друг от друга отличаются?

ErazorWhite replied:
ResultsBoxCalculated это общий контейнер для поля Results. Переименную в ResultsCalculatedContainer, чтобы дать понять, что он отвечает за расположение вложенных блоков.
CalculatedBox - изменю имя на CalculatedNumbersBox, чтобы было понятно, что это более мелкий вложенный блок, который содержит в себе цифры расчётов.

afitiskin commented on src/components/Results/CalculatedResults/CalculatedResults.tsx:
```diff
@@ -0,0 +1,23 @@
+        <ResultsBoxCalculated>
+            <H2>Your results</H2>
+            <Description>
+                Your results are shown below based on the information you provided. To adjust the results, edit
```
Я бы дал возможность этот текст задать снаружи, и зафолбечился бы на этот "базовый"

ErazorWhite replied:
Было бы не плохо в константы вынести куда-то, пожалуй.
Но пока так:
```
export const CalculatedResults = ({
                                      monthlyRepayment,
                                      totalRepayment,
                                      currency = "GBP",
                                      calculatedHeadingText = "Your results",
                                      calculatedDescriptionText = "Your results are shown below based on the information you provided. To adjust the results, edit\n" +
                                      "                the form and click “calculate repayments” again.",
                                      numbersPrimaryLabelText = "Your monthly repayments",
                                      numbersSecondaryLabelText = "Total you'll repay over the term",
                                  }: ResultsProps) => {
render
}
```

afitiskin commented on src/components/Results/CalculatedResults/CalculatedResults.tsx:
```diff
@@ -0,0 +1,23 @@
+            </Description>
+
+            <CalculatedBox>
+                <P>Your monthly repayments</P>
```
Тоже можно сделать параметром

ErazorWhite replied:
Done

afitiskin commented on src/components/Results/CalculatedResults/CalculatedResults.tsx:
```diff
@@ -0,0 +1,23 @@
+
+            <CalculatedBox>
+                <P>Your monthly repayments</P>
+                <H3>£{Intl.NumberFormat('en-US').format(Number(monthlyRepayment))}</H3>
```
Во-первых, NumberFormat в том числе придуман для того, чтобы правильно и в нужном месте вставлять знак валюты. Во-вторых, сама валюта должна задаваться параметром вместе с числами. Ну и приведение к Number выглядит странно, почему не получить число сразу извне?

afitiskin replied:
Посмотрел тип, у тебя там number | null. Но если тебе пришел null, что ты можешь показать? Наверное ничего. Поэтому не вижу смысла тут принимать null вообще

afitiskin commented on src/components/Results/CalculatedResults/CalculatedResults.tsx:
```diff
@@ -0,0 +1,23 @@
+import React from 'react'
+import {ResultsProps} from "../../../global/types";
```
Отрывать тип от компонента плохая идея. Если тип небольшой – можно описать его прямо тут, если большой, в соседнем файле но в этой же папке. Крайний случай папка types в этой папке

ErazorWhite replied:
Прошелся по компонентам, передислоцировал типизацию согласно этой рекомендации

afitiskin commented on src/components/Results/CalculatedResults/CalculatedResults.tsx:
```diff
@@ -0,0 +1,23 @@
+                <H3>£{Intl.NumberFormat('en-US').format(Number(monthlyRepayment))}</H3>
+
+                <P>Total you'll repay over the term</P>
+                <H4>£{Intl.NumberFormat('en-US').format(Number(totalRepayment))}</H4>
```
У тебя на каждый рендер по 2 раза создается новый объект number format.

При инициализации компоненты или при смене валюты создаешь новый экземпляр форматтера, мемоизируешь его и используешь во всех рендерах. При желании (если он еще где-то используется) можно вообще 1 экземпляр держать в памяти и использовать его через контекст во всем приложении

ErazorWhite replied:
null выпилил
форматтер мемоизировал
```
    const formatter = useMemo(
        () =>
            new Intl.NumberFormat('en-US', {
                style: 'currency',
                currency: currency,
            }),
        [currency]
    );
```

```
<NumbersLabel>{numbersPrimaryLabelText}</NumbersLabel>
                <NumbersHeading primary>{formatter.format(monthlyRepayment)}</NumbersHeading>

                <NumbersLabel>{numbersSecondaryLabelText}</NumbersLabel>
                <NumbersHeading>{formatter.format(totalRepayment)}</NumbersHeading>
```

afitiskin commented on src/global/theme.ts:
```diff
@@ -0,0 +1,39 @@
+        font-size: 56px;
+        line-height: 1.25;
+        letter-spacing: -1px;
+        font-family: 'Plus Jakarta Sans', serif, Roboto;
```
робото у тебя никогда не применится, потому что serif это полный fallback. Робото и Plus Jakarta Sans (тут даже по названию) – шрифты БЕЗ засечек. У тебя фоллбек на шрифты с засечками, почему?

ErazorWhite replied:
Roboto вынес на второе место
serif поменял на sans-serif и оставил третьим на очереди

afitiskin commented on src/index.css:
```diff
@@ -1,68 +1,81 @@
-  transition: border-color 0.25s;
+@font-face {
+    font-family: 'Plus Jakarta Sans';
+    src: url('./assets/fonts/PlusJakartaSans-Bold.woff2'),
```
указывай формат шрифта, чтобы браузер еще до загрузки файла понимал, поддерживает он этот формат или нет (https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face)

ErazorWhite replied:
Поправил

afitiskin commented on src/global/theme.ts:
```diff
@@ -0,0 +1,39 @@
+import {css} from "styled-components";
+
+export const typography = {
+    preset1: css`
```
не скейлится такой подход. ты уже через неделю сам забудешь что такое preset1, а что такое preset3

ErazorWhite replied:
Я все делаю по макету, пресеты взяты оттуда.
Дизайнер нарисовал текстовый элемент и задал ему text-preset-4.
Если я названиям шрифтов задам свои собственные обозначения, то на макет через неделю буду смотреть как баран на новые ворота.
![image](https://github.com/user-attachments/assets/829eb27b-b532-40d3-b95d-915dc355c36f)
![image](https://github.com/user-attachments/assets/1607d326-6afb-43d5-b451-0250c3c6c039)

afitiskin replied:
Ок, тогда по рукам такому дизайнеру нужно бить

afitiskin commented on src/components/ValidateError/ValidateError.styled.ts:
```diff
@@ -0,0 +1,8 @@
+import styled from "styled-components";
+import {typography} from "../../global/theme";
+
+export const ValidateErrorStyled = styled.p`
```
ошибка это уж точно не абзац

afitiskin commented on src/components/utilities/mortgageCalc.ts:
```diff
@@ -0,0 +1,21 @@
+import {MortgageType} from "../../global/types.ts";
+
+export function mortgageCalc(term: number, rate: number, type: MortgageType, amount: number) {
```
на память назовешь порядок параметров?

ErazorWhite replied:
здесь лучше сгруппировать в один объект

afitiskin commented on src/components/utilities/mortgageCalc.ts:
```diff
@@ -0,0 +1,21 @@
+    const mortgageTermMonths = term * 12;
+    const monthlyInterestRate = (rate / 100) / 12;
+
+    let monthlyPayment: number;
```
не вижу смысла в let если в каждом отдельном кейсе можно вернуть нужный объект

ErazorWhite replied:
Можно сделать так. let не будет
```
  if (type === MortgageType.Repayment) {
    const numerator =
      amount *
      monthlyInterestRate *
      Math.pow(1 + monthlyInterestRate, mortgageTermMonths);
    const denominator =
      Math.pow(1 + monthlyInterestRate, mortgageTermMonths) - 1;

    const monthlyPayment = numerator / denominator;
    const totalRepayment = monthlyPayment * mortgageTermMonths;

    return { monthlyPayment, totalRepayment };
  } else {
    const monthlyPayment = amount * monthlyInterestRate;
    const totalRepayment = monthlyPayment * mortgageTermMonths;

    return { monthlyPayment, totalRepayment };
  }

```

afitiskin commented on src/global/types.ts:
```diff
@@ -0,0 +1,66 @@
+
+// Text input
+
+export enum Position {
```
ну это точно не тип, а вполне конкретные значения как и MortgageType

ErazorWhite replied:
Пока дошел до этой правки, то Position полностью выпилил. Так что тут править нечего.

afitiskin commented on src/components/form/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,38 @@
+import {InputProps} from "../../../global/types";
+import {ValidateErrorStyled} from "../../ValidateError/ValidateError.styled";
+
+export const TextInput = ({name, label, control, prefix, suffix}: InputProps) =>
```
ну это точно никак не TextInput

ErazorWhite replied:
Поменяю на FormattedNumberInput

afitiskin commented on src/components/form/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,38 @@
+                    <TextLabel htmlFor={name}>{label}</TextLabel>
+                    <InputBox data-prefix={prefix} data-suffix={suffix} $isInvalid={!!error} $isTouched={isTouched && !error}>
+
+                        <NumericFormat
```
Рассчитывал все-таки увидеть кастом, но увы :(

ErazorWhite replied:
Я несколько часов попытался, потом появилось ощущение, что много времени трачу на создание кастома. Решил сэкономить время, но пожертвовать 6.5 kB. Думаю важнее добить челлендж до конца, а задачу переделать это в кастом можно на потом оставить.

afitiskin commented on src/components/form/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,38 @@
+            render={({field: {onChange, onBlur, value, ref}, fieldState: {error, isTouched}}) => (
+                <>
+                    <TextLabel htmlFor={name}>{label}</TextLabel>
+                    <InputBox data-prefix={prefix} data-suffix={suffix} $isInvalid={!!error} $isTouched={isTouched && !error}>
```
давай не будем изобретать $-нотацию, мы не в php

ErazorWhite replied:
Если не добавлять $-префикс к пропсам, которые я передаю в styled,  в консоли React будет ругаться.

```
Warning: React does not recognize the stylesForEmpty prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase stylesforempty instead. If you accidentally passed it from a parent component, remove it from the DOM element.
```

Ошибка говорит, что React не знает, что делать с пропсом, который передаётся напрямую в DOM-элемент. Пропсы с $-префиксом это транзитные пропсы для styled-components, они не передаются в DOM-элемент, но остаются доступными внутри styled-components для стилизации.

Если не хочешь видеть $-префиксы, то можем использовать из-под капота styled вот такую штуку.

```
export const ResultsHeading = styled.h2.withConfig({
  shouldForwardProp: (prop) => prop !== 'stylesForEmpty',
})<ResultsHeadingProps>`
  ${({ stylesForEmpty }) => stylesForEmpty ? css`
    margin: 16px 0;
  ` : css`
    margin: 0 0 16px 0;
  `}
`;
```

Здесь я привел решения на примере компонента Results, над улучшением которого я сейчас работаю, но эти же выводы можно экстраполировать и на InputBox, о котором идёт выше речь.

ErazorWhite replied:
Здесь в итоге ушел от $-нотации при помощи 
```
.withConfig({ shouldForwardProp: (prop) => prop !== 'propName', }
```

afitiskin commented on src/components/form/TextInput/TextInput.styled.ts:
```diff
@@ -0,0 +1,62 @@
+    color: var(--color-state-700);
+`
+
+export const InputBox = styled.div<InputBoxProps>`
```
если ты используешь data- параметры, то они обязаны быть задекларированы в типах

ErazorWhite replied:
done
```
interface InputBoxProps {
    isInvalid?: boolean;
    isTouched?: boolean;
    'data-prefix'?: string;
    'data-suffix'?: string;
}
```

afitiskin commented on src/components/form/TextInput/TextInput.tsx:
```diff
@@ -0,0 +1,38 @@
+
+                    </InputBox>
+                    {error && <ValidateErrorStyled>{error.message}
+                    </ValidateErrorStyled>
```
тут явно что-то с форматированием

afitiskin commented on src/components/form/RadioGroup/RadioGroup.styles.ts:
```diff
@@ -0,0 +1,80 @@
+    margin-bottom: 12px;
+`
+
+export const RadioLabel = styled.label`
```
Ну давай так, стили радио-элементов не привязаны к форме напрямую и могут использоваться в других местах приложения, например в простом диалоге, где не используется rhf а используется внутренний стейт. Поэтому было бы неплохо четко отделить, что ты можешь переиспользовать без rhf, а что используется только с rhf и без него не работат

ErazorWhite replied:
Идею уловил, но что конкретно сейчас сделать не совсем понимаю. 
Стили лейблов и вот эти ::before ::after с :checked вынести куда-то отдельно от компонента?
Сейчас это не имеет практического применения. Возможно в дальнейшем поможет, но я не вижу здесь конкретной цели, чтобы сесть и сделать что ты хочешь.

afitiskin replied:
Ну мы в прошлом челлендже уже затрагивали эту тему, тут снова сталкиваемся с подобным..

ErazorWhite replied:
Всё, до меня дошло что нужно
Теперь должно быть красиво [RadioInput extracted from RadioGroup](https://github.com/ErazorWhite/mortgage-repayment-calculator/pull/1/commits/8cdfad097100f14cd82e90debdb37b758e686a99)

afitiskin commented on src/global/types.ts:
```diff
@@ -0,0 +1,66 @@
+
+export interface RadioGroupProps {
+    label: string;
+    name: keyof MortgageData;
```
странное ограничение

ErazorWhite replied:
согласен, тут больше пойдёт string

ErazorWhite replied:
Одним string тут оказалось никак не отделаться.
name: `Path<TFieldValues>` гарантирует, что name соответствует существующему полю в данных формы.
Но и типизацию register понадобилось отвязать от MortgageData. 
Короче опять я гвоздями всё поприбивал, теперь с фомкой λ работаю в обратном направлении.
```
export interface RadioGroupProps<TFieldValues extends FieldValues> {
    label: string;
    name: Path<TFieldValues>;
    options: RadioOption[];
    register: UseFormRegister<TFieldValues>;
    error?: FieldError;
}
```
И сигнатурка компонента тоже поменялась
```
export const RadioGroup = <TFieldValues extends FieldValues, >({
                                                                   label,
                                                                   name,
                                                                   options,
                                                                   register,
                                                                   error
                                                               }: RadioGroupProps<TFieldValues>) => {
```

afitiskin commented on src/components/form/RadioGroup/RadioGroup.styles.ts:
```diff
@@ -0,0 +1,80 @@
+    width: 0;
+
+    &:focus + ${RadioLabel} {
+        outline: 2px solid black;
```
черый нужно добавить в палитру (потому что может быть оттенок) и используй цвета всегда в одном формате, не смешивай названия цветов, hex, rbg, hsl и т.п.

ErazorWhite replied:
done

afitiskin commented on src/components/form/RadioGroup/RadioGroup.styles.ts:
```diff
@@ -0,0 +1,80 @@
+import {typography} from "../../../global/theme";
+import {TABLET_BP} from "../../../global/breakpoints";
+
+export const StyledRadio = styled.div`
```
Если у тебя заголок `legend`, то тут должен быть `fieldset`, иначе использовать `legend` нельзя

ErazorWhite replied:
fixed

afitiskin commented on src/components/form/CalculatorForm/CalculatorForm.tsx:
```diff
@@ -0,0 +1,78 @@
+                    <Li>
+                        <TextInput name="amount" label="Mortgage Amount" prefix="£" control={control}/>
+                    </Li>
+                    <Li>
```
Из кода никак не следует, что это и следующее поле в 50% ширины, в отличии от первого. Если мы захотим порядок полей местами поменять, твоя верстка нам этого не позводит. Получается очень негибко

ErazorWhite replied:
сделал FullLi, HalfLi, которые наследуют общие свойства от Li, но отличаются шириной на TABLET_BP+

afitiskin commented on src/components/form/CalculatorForm/CalculatorForm.tsx:
```diff
@@ -0,0 +1,78 @@
+        formState: {errors}
+    } = useForm<MortgageData>({resolver: yupResolver(MortgageDataSchema), mode: "onTouched"});
+
+    const onSubmit: SubmitHandler<MortgageData> = ({amount, term, rate, type}) => {
```
мемоизация (в принципе забыта в этом проекте)

ErazorWhite replied:
добавил мемоизации

scd-hoi committed:
review changes

favicon rename, remove unnecessary stuff, button radius fix, add button props, RadioGroupProps type fix, legend fix, split Li to full/half

scd-hoi committed:
review changes 2

remove Results.tsx,
modify Calculated & Empty Results,
add text parameters
rework styles to be more generic,
remove $-notation,
add "data-" types,
keep color vars consistent,
remove global/types, add types directly,
serif font issue fix

NotAleksei commented on src/components/CalculateButton/CalculateButton.styled.ts:
```diff
@@ -0,0 +1,22 @@
+    transition: var(--animation);
+
+    &:hover,
+    &focus {
```
Псевдокласс `:focus` без двоеточий не будет работать.

ErazorWhite replied:
fixed

NotAleksei commented on src/components/Results/CalculatedResults/CalculatedResults.tsx:
```diff
@@ -0,0 +1,60 @@
+                {calculatedDescriptionText}
+            </Description>
+
+            <CalculatedNumbersBox>
```
Честно говоря и `ResultsCalculatedContainer` и `CalculatedNumbersBox` - бросаются в глаза. Очень общие слова для конкретной имплементации mortgage calculator.
 `ResultsCalculatedContainer` и само название компонента `CalculatedResults` как минимум неоправданно отличаются порядком слов. Ну и я бы больше предпочел существительное как `CalculationResult`. 
 
Что касается Numbers, то это очень дженерик слово, да и в целом дальше у тебя не numbers а текст с информацией и результатом вычислений. Если мы хотим все же придерживаться общих слов, без уточнения зачем этот компонент нужен, то я бы остановился на CalculationDetails и дальше уже можно добавлять приписки Box/Label etc.

ErazorWhite replied:
Тут в целом прошелся по всем компонентам в папке Results и подправил названия то тут, то там.

NotAleksei commented on src/components/Results/CalculatedResults/CalculatedResults.tsx:
```diff
@@ -0,0 +1,60 @@
+            </Description>
+
+            <CalculatedNumbersBox>
+                <NumbersLabel>{numbersPrimaryLabelText}</NumbersLabel>
```
`NumbersLabel` и `NumbersHeading` - label говорит больше о назначении чем о теге, а heading - описывает просто стилизованный h3 тег.

Да и само исползование h3 тега тут семантически не очень корректно. Все же он больше подразумевает разделение информации на какие-то главы. А тут используется просто чтобы выделить результат вычисления. Стилизованные `<span>` или `<p>` были бы более ожидаемым.

Сам лейбл вполне мог бы ожидаться h3 по структуре, хотя я бы предпочел обычный `span` судя по содержимому.

ErazorWhite replied:
Теперь так

`CalculationLabel` - содержит название поля, использует тег `p`
`CalculationValue` - содержит результат вычисления, использует тег `span`

Я использовал разные теги т.к. по макету лейбл менее взрачный, серый и блеклый. Сюда кастанул `p`. Любые `p` внутри родительского контейнера будут иметь одинаковый пресет из объекта типографии и так же одинаковый цвет. 

Под результат вычисления дизайнер отвёл бОльшую палитру стилей, потому я решил использовать `span`, чтобы выделить этот фрагмент текста.

NotAleksei commented on src/components/Results/EmptyResults/EmptyResults.tsx:
```diff
@@ -0,0 +1,21 @@
+                             }: EmptyResultsProps) => (
+    <ResultsBoxEmptyContainer>
+        <ResultsBackground/>
+        <ResultsHeading stylesForEmpty>{emptyResultsHeadingText}</ResultsHeading>
```
`stylesForEmpty` на первый взгляд подразумевает что в пропертю надо положить какие-то стили. Хотя по своей сути является просто булевым значением, которое должно описывать `ResultsHeading`. Я бы предложил использовать нейминг как `isEmpty` или как возможный вариант просто `empty`.

ErazorWhite replied:
Плюсую, поправил

NotAleksei commented on src/components/ValidateError/ValidationErrorStyled.ts:
```diff
@@ -0,0 +1,8 @@
+
+export const ValidationErrorStyled = styled.p`
+    color: var(--color-error); 
+    ${typography.preset5};
```
Стоит указать preset перед color, т.к. если вдруг в preset появится color свойство, то оно перетрет текущее значение `var(--color-error)`.

ErazorWhite replied:
done

NotAleksei commented on src/components/form/CalculatorForm/CalculatorForm.styled.ts:
```diff
@@ -0,0 +1,84 @@
+        display: flex;
+    }
+
+    box-shadow: 0 32px 64px rgba(19, 48, 65, 0.1);
```
Тут субъективно, но я бы располагал медиа в самом конце стилей.

ErazorWhite replied:
Почему бы и нет, медиа довольно массивные, пусть будут расположены в своей нише.

NotAleksei commented on src/components/form/CalculatorForm/CalculatorForm.styled.ts:
```diff
@@ -0,0 +1,84 @@
+
+export const H1 = styled.h1`
+    margin-bottom: 8px;
+    ${typography.preset2};
```
Аналогично, если не преследуешь цели чтобы пресет перетирал `margin-bottom`, то лучше расположить его повыше.

ErazorWhite replied:
Плюс

NotAleksei commented on src/components/form/CalculatorForm/CalculatorForm.styled.ts:
```diff
@@ -0,0 +1,84 @@
+
+export const Ul = styled.ul`
+    margin: 0 0 24px 0;
+    flex-wrap: wrap;
```
`flex-wrap` не сработает, т.к. элемент не имеет `flex` проперти

ErazorWhite replied:
В целом роли не играет, свойство `flex-wrap` мне нужно в медиа `TABLET_BP`, где уже прописан `display: flex`. Тут, чтобы было правильно и красиво, закинул wrap в медиа.

NotAleksei commented on src/components/form/CalculatorForm/CalculatorForm.tsx:
```diff
@@ -0,0 +1,88 @@
+                    </ClearButton>
+                </FormHeader>
+
+                <Ul>
```
Использовать `ul` (unordered list) и `li` (list item) теги для организации текстовых полей в форме - семантически некорректно. Они применяются для организации неупорядоченных списков. И люди которые пользуются скринридерами будут ожидать список каких-то элементов, а не набор полей.

Допускается использовать их в форме для списка чекбоксов или радио кнопок см. рекомендации [MDN](https://developer.mozilla.org/en-US/docs/Learn/Forms/How_to_structure_a_web_form#common_html_structures_used_with_forms)

Если надо сгруппировать какие-то поля - используй `fieldset` или просто `div`

ErazorWhite replied:
Уф, а меня учили, что любые повторяемые элементы страницы следует прятать под `ul > li`. 
Но developer.mozilla.org, конечно, будет авторитетнее в этом вопросе. 

Тут поменял всё на div и изменил имена.

NotAleksei commented on src/components/form/CalculatorForm/CalculatorForm.tsx:
```diff
@@ -0,0 +1,88 @@
+
+const radioOptions = [
+    {label: MortgageType.Repayment, value: MortgageType.Repayment},
+    {label: MortgageType.Repayment, value: MortgageType.InterestOnly},
```
Лейблы одинаковые для разных опций

ErazorWhite replied:
fixed

NotAleksei commented on src/components/form/CalculatorForm/CalculatorForm.tsx:
```diff
@@ -0,0 +1,88 @@
+
+    const onSubmit: SubmitHandler<MortgageData> = useCallback(({amount, term, rate, type}) => {
+
+        let {monthlyPayment, totalRepayment} = mortgageCalc({term, rate, type, amount});
```
`const` вместо `let`. Но обычно линтер такое правит.

ErazorWhite replied:
fixed

NotAleksei commented on src/components/Results/ResultsCommon.styled.ts:
```diff
@@ -0,0 +1,46 @@
+    }
+`
+export const ResultsHeading = styled.h2.withConfig({ shouldForwardProp: (prop) => prop !== 'stylesForEmpty', })<ResultsHeadingProps>`
+    ${({stylesForEmpty})=> stylesForEmpty ? css`
```
Тут выглядит немного избыточно.

Будет достаточно:

```css
margin: ${({stylesForEmpty})=> stylesForEmpty ? `16px 0;` : `0 0 16px 0;`}
```

ErazorWhite replied:
Эх, а ведь интуиция подсказывала мне, что конструкция слишком громоздкая, да и в [styled docs](https://styled-components.com/docs/basics#passed-props) есть примерчики. 
Спасибо.

NotAleksei commented on src/components/form/CalculatorForm/CalculatorForm.tsx:
```diff
@@ -0,0 +1,88 @@
+        reset();
+    }, []);
+
+    const onError: SubmitErrorHandler<MortgageData> = useCallback((data) => console.log(data), []);
```
Кажется что console.log тут лишний (если, конечно, это не мок какого-то другого поведения в рамках обучения)

ErazorWhite replied:
Выпилено

NotAleksei commented on src/components/form/CalculatorForm/CalculatorForm.tsx:
```diff
@@ -0,0 +1,88 @@
+
+                <CalculateButton type="submit">Calculate Repayments</CalculateButton>
+
+            </Form>
```
Лишний пробел.
Вообще, я бы поубирал пробелы между блоками тут везде. Но на твое усмотрение

ErazorWhite replied:
Мне в целом ок с отступами между блоками/тегами, меньше глаза напрягаются Остальное prettier сам правит.
Тут пробелов уменьшил.

NotAleksei commented on src/components/form/FormattedNumberInput/FormattedNumberInput.styled.ts:
```diff
@@ -0,0 +1,69 @@
+    color: var(--color-state-700);
+`
+
+export const InputBox = styled.div.withConfig({shouldForwardProp: (prop) => prop != 'isInvalid' && prop != 'isTouched'})<InputBoxProps>`
```
Используй строгие операторы сравнения, пожалуйста.

ErazorWhite replied:
done

NotAleksei commented on src/components/form/FormattedNumberInput/FormattedNumberInput.tsx:
```diff
@@ -0,0 +1,46 @@
+                            decimalScale={2}
+                            value={value ?? ''}
+                            onBlur={onBlur}
+                            onValueChange={useCallback((values: NumberFormatValues) => {
```
Ты должен использовать хуки на верхнем уровне своего компонента см [доку](https://react.dev/reference/react/useCallback#caveats)

Конкретно в этом случае я бы не стал сильно заморачиваться с оборачиванием в `useCallback`, предположу что компонент все равно будет перерисовываться каждый раз когда меняется value из аргументов функции render

ErazorWhite replied:
Убрал useCallback

NotAleksei commented on src/components/form/RadioGroup/RadioGroup.styles.ts:
```diff
@@ -0,0 +1,80 @@
+    }
+}
+`;
+
```
Тут лишние переносы строки в конце файла

ErazorWhite replied:
Тут перевернулась фура с переносами 😓

NotAleksei commented on src/components/utilities/mortgageCalc.ts:
```diff
@@ -0,0 +1,21 @@
+    const monthlyInterestRate = (rate / 100) / 12;
+
+    if (type === MortgageType.Repayment) {
+        const numerator = amount * monthlyInterestRate * Math.pow(1 + monthlyInterestRate, mortgageTermMonths);
```
Возможно тебе стоит добавить валидацию на возможный ввод поля term и rate. Т.к. при больших значениях (оба больше 1000 скажем) - ты будешь получать тут Infinity а форма никак не среагирует на это.

ErazorWhite replied:
Запомню на будущее, что с Math.pow надо быть аккуратнее.
Закинул в yup .max() валидацию.

NotAleksei commented on src/components/Results/CalculatedResults/CalculatedResults.tsx:
```diff
@@ -0,0 +1,60 @@
+    numbersSecondaryLabelText?: string;
+}
+
+export const CalculatedResults = ({
```
Если ты начнешь вводить большие числа то форма разъедется. Стоит поправить стили чтобы предусмотреть это. На мобилке тоже.
![image](https://github.com/user-attachments/assets/ea64fd3c-6cfc-4ce0-9b98-b38bbb838180)

ErazorWhite replied:
На NumbersHeading добавил overflow: scroll. Компоненту CalculatedResults добавил фиксированную ширину 50% от родителя при DESKTOP_BP

NotAleksei commented on src/components/form/FormattedNumberInput/FormattedNumberInput.styled.ts:
```diff
@@ -0,0 +1,69 @@
+`
+
+export const StyledInput = styled.input`
+    padding: 13px 16px;
```
У тебя сейчас инпут вылезает за границы скругленного родителя. Отнаследуй радиус чтобы избежать этого
`border-radius: inherit;`

ErazorWhite replied:
Если бы ты не сказал, то я бы и не заметил.
Глаз-алмаз :not-bad:
![image](https://github.com/user-attachments/assets/93efd0db-e607-4c59-baa1-6382bb9eb9b6)
![image](https://github.com/user-attachments/assets/bf06b855-ce95-4e76-814d-6781528e392a)

NotAleksei commented:
Я бы может еще пересмотрел структуру папок.
У тебя utilities лежит в components - можно вынести выше.

Дальше обсуждаемо, но я бы разделял контейнеры и компоненты. Компоненты - как конечные небольшие элементы, которые ты переиспользуешь в проекте (кнопки/инпуты/ошибки). Контейнеры - как обертки над группой элементов - как форма или блок Results

ErazorWhite commented:
> Я бы может еще пересмотрел структуру папок. У тебя utilities лежит в components - можно вынести выше.

Done
 
> Дальше обсуждаемо, но я бы разделял контейнеры и компоненты. Компоненты - как конечные небольшие элементы, которые ты переиспользуешь в проекте (кнопки/инпуты/ошибки). Контейнеры - как обертки над группой элементов - как форма или блок Results

Предложение дельное, хотя это больше проблемы будущего меня в случае расширения проекта. Если я начну делать огромный проект, в котором эта часть формы (контейнер Results и подобные ему блоки) повторно используется ещё в 20 местах, то я безусловно устану рисовать один и тот же контейнер снова и снова. Когда я столкнусь с этим, то точно задумаюсь над решением. А пока передо мной эта проблема остро не стоит, то это что-то абстрактное гипотетическое, над решением которого думать сейчас может быть не очень целесообразно. Я пока предпочёл бы сконцентрироваться на проблемах третьего челленджа. 

Хотя декомпозировать CalculatorForm на ещё один подкомпонент CalculatorFormFields было бы не плохо уже сейчас. 
Что-то подобное я уже практиковал с формиком здесь: https://github.com/ErazorWhite/contact-form/tree/main/src/components/form

scd-hoi committed:
review changes 3

move utilities out of components,
validation term/rate .max fix,
typography presets moved on top,
results overflow fix,
stylesForEmpty renamed to isEmpty,
make styled props less redundant,
remove redundant useCallback,
use strict comparisons,
rename result components,
correct radioOptions label,
flex-wrap moved into @media,
box-shadow moved on top,
&focus fix,
change ul>li to div for inputs,
remove extra spaces,

afitiskin commented on src/components/Results/CalculationResult/CalculationResult.tsx:
```diff
@@ -0,0 +1,60 @@
+export const CalculationResult = ({
+                                      monthlyRepayment,
+                                      totalRepayment,
+                                      currency = "GBP",
```
У тебя кавычки почему то одновременно и одинарные и двойные используются. Нужно определиться и использовать либо одни либо другие

ErazorWhite replied:
Поменял везде на одинарные.
Сначала всё сломалось ибо в `CalculationResult` был пропс `calculationSecondaryLabelText = 'Total you'll repay over the term',`
Тут можно обратный слешик поставить `\'`.
Но решил отказаться от сокращений, сделаю вид, что это очень деловой диалог.

afitiskin commented on src/components/Results/EmptyResult/EmptyResult.tsx:
```diff
@@ -0,0 +1,21 @@
+import {ResultsBackground} from "../../icons/ResultsBackground.tsx";
+import {ResultsBoxEmptyContainer} from "./EmptyResult.styled.ts";
+
+interface EmptyResultsProps {
```
Типам желательно задавать префикс I чтобы можно было отличать их от самих компонентов и констант

ErazorWhite replied:
done

afitiskin commented on src/components/Results/ResultsCommon.styled.ts:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    p {
+        ${typography.preset4};
```
Вот эти стили нужно вынести на уровень выше. Если внутри блока по какой-то причине мы не сможем использовать абзац а например используем что-то другое, то нужно чтобы текст выглядел правильно

ErazorWhite replied:
Ок, закинул просто на ResultsBox styled.div

afitiskin commented on src/components/Results/ResultsCommon.styled.ts:
```diff
@@ -0,0 +1,43 @@
+    }
+`
+export const Heading = styled.h2.withConfig({ shouldForwardProp: (prop) => prop !== 'isEmpty', })<ResultsHeadingProps>`
+    margin: ${({isEmpty}) => isEmpty ? `16px 0;` : `0 0 16px 0;`}
```
Точку с запятой можно оставить в конце 1 раз

ErazorWhite replied:
done

afitiskin commented on src/components/form/CalculatorForm/CalculatorForm.tsx:
```diff
@@ -0,0 +1,80 @@
+
+        const {monthlyPayment, totalRepayment} = mortgageCalc({term, rate, type, amount});
+
+        setMonthlyRepayment(parseFloat(monthlyPayment.toFixed(2)));
```
Зачем кастовать строку если можно математически округлить?

ErazorWhite replied:
Marh.round((num * 100)) / 100 сделал вместо toFixed(). И выкинул это из компонента формы под капот функции calculateMortgage

afitiskin commented on src/components/form/CalculatorForm/types.ts:
```diff
@@ -0,0 +1,11 @@
+export enum MortgageType {
```
Опять у тебя данные лежат в файле с типами

ErazorWhite replied:
так как этот enum используется в четырёх разных файлах (types.ts, CalculatorForm.tsx, calculateMortgage.ts, validationScemas.ts), то я его выкинул в ~/global/sharedData.ts

afitiskin commented on src/components/form/FormattedNumberInput/FormattedNumberInput.styled.ts:
```diff
@@ -0,0 +1,70 @@
+    color: var(--color-state-700);
+`
+
+export const InputBox = styled.div.withConfig({shouldForwardProp: (prop) => prop !== 'isInvalid' && prop !== 'isTouched'})<InputBoxProps>`
```
[].includes()

ErazorWhite replied:
done

afitiskin commented on src/components/form/RadioGroup/RadioGroup.styles.ts:
```diff
@@ -0,0 +1,74 @@
+    }
+`
+
+export const RadioHeader = styled.p`
```
Меня все еще смущает использование тега p для чего либо кроме абзаца текста

ErazorWhite replied:
Я всё таки решил применить здесь fieldset + legend.
```
StyledRadio fieldset 
> RadioHeader legend
> ul
>> Li
>>> RadioInput
>>> RadioLabel
```

afitiskin commented on src/utilities/mortgageCalc.ts:
```diff
@@ -0,0 +1,21 @@
+import {MortgageData, MortgageType} from "../components/form/CalculatorForm/types.ts";
+
+export function mortgageCalc({term, rate, type, amount}: MortgageData) {
```
Функция это действие, отсюда название должно быть глаголом.

ErazorWhite replied:
Имя изменено

scd-hoi committed:
change result calculation round logic

scd-hoi committed:
RadioInput extracted from RadioGroup

scd-hoi committed:
"" changed to '',
start interface names from I,
MortgageType moved to /global/sharedData,
ResultsBox p styles moved on top,

scd-hoi committed:
make ref optional

kpogodin commented:
Предлагаю донастроить конфиг eslint. Файлы выглядят сильно непричесаными, со странной табуляцией, ide ничего не подсвечивает. 

1. устанавливаем дополнительно `eslint-plugin-prettier ``prettier` `typescript-eslint.`
2. апгрейдим конфиг <img width="824" alt="Screenshot 2024-10-29 at 4 02 53 PM" src="https://github.com/user-attachments/assets/806fb369-c287-40f7-a0cf-b33106776d60">
3. Получаем сочные полностью красные файлы. <img width="1069" alt="Screenshot 2024-10-29 at 4 01 29 PM" src="https://github.com/user-attachments/assets/3ab4868f-6879-4a76-8de4-b7689dfb7844">
4. Устанавливаем экстеншн Prettier и выставляем в настройках IDE автопочинку файла при сохранении (в VSCode в settings -> `"editor.formatOnSave": true`)
<img width="360" alt="Screenshot 2024-10-29 at 4 19 33 PM" src="https://github.com/user-attachments/assets/45043a1d-5508-469d-af31-e3cfdfd7ab97">

5. добавляем файл`.prettierrc.cjs` например с таким конфигом (как у нас в БО):  
```
const config = {
  arrowParens: 'always',
  bracketSameLine: false,
  endOfLine: 'auto',
  jsxSingleQuote: false,
  printWidth: 100,
  semi: true,
  singleQuote: true,
  tabWidth: 2,
  trailingComma: 'es5',
};

module.exports = config;
```

6. Получаем вот такой красивый файл при его сохранении 
<img width="921" alt="Screenshot 2024-10-29 at 4 01 35 PM" src="https://github.com/user-attachments/assets/ee51e356-3fad-417b-932e-cca113f27d68">

kpogodin commented on src/index.css:
```diff
@@ -1,68 +1,86 @@
+    --color-black: black;
+
+    /* Animation */
+    --animation: color 150ms ease-in, background-color 150ms ease-in, border-color 150ms ease-in;
```
Стоит дать более точно название. Что будет если появится вторая анимация?)

ErazorWhite replied:
done

kpogodin commented on src/components/form/FormattedNumberInput/FormattedNumberInput.tsx:
```diff
@@ -0,0 +1,45 @@
+import {InputBox, StyledInput, TextLabel} from './FormattedNumberInput.styled.ts';
+import {NumberFormatValues, NumericFormat} from 'react-number-format';
+import {Control, Controller} from 'react-hook-form';
+import {ValidationErrorStyled} from '../../ValidateError/ValidationErrorStyled.ts';
```
файлы со styled компонентами иногда называются `*.styled.ts`, а иногда прям слитно как тут `ValidationErrorStyled.ts`

ErazorWhite replied:
это по запаре
как правило везде `*.styled.ts`
поправил

kpogodin commented on src/main.styled.ts:
```diff
@@ -0,0 +1,23 @@
+
+    @media screen and ${TABLET_BP} {
+        max-width: 768px;
+        padding: 40px 40px 40px 40px;
```
padding: 40px

ErazorWhite replied:
fixed

scd-hoi committed:
update and use linter & prettier

scd-hoi committed:
README update
